Content.Shared/Chemistry/Components/ReagentDispenserComponent.cs

```diff
@@ -0,0 +1,105 @@
+using Content.Shared.Chemistry.EntitySystems;
+using Content.Shared.Containers.ItemSlots;
+using Content.Shared.FixedPoint;
+using Content.Shared.Storage;
+using Robust.Shared.Audio;
+using Robust.Shared.GameStates;
+using Robust.Shared.Serialization;
+
+namespace Content.Shared.Chemistry.Components;
+
+/// <summary>
+/// A machine that dispenses reagents into a solution container from containers in its storage slots.
+/// </summary>
+[RegisterComponent, NetworkedComponent, AutoGenerateComponentState]
+[Access(typeof(ReagentDispenserSystem))]
+public sealed partial class ReagentDispenserComponent : Component
+{
+    public const string OutputSlotName = "beakerSlot";
+
+    [DataField]
+    public ItemSlot BeakerSlot = new();
+
+    [DataField]
+    public SoundSpecifier ClickSound =
+        new SoundPathSpecifier("/Audio/Machines/machine_switch.ogg", AudioParams.Default.WithVolume(-2f));
+
+    [DataField, AutoNetworkedField]
+    public FixedPoint2 DispenseAmount = 10;
+
+    // Collection expression is a sandbox violation here. :(
+    // ReSharper disable once UseCollectionExpression
+    /// <summary>
+    /// What amount options to show in the reagent amount selector... thing.
+    /// </summary>
+    [DataField]
+    public List<FixedPoint2> SelectableAmounts = new() { 1, 5, 10, 15, 20, 25, 30, 50, 100 };
+}
+
+[Serializable, NetSerializable]
+public sealed class ReagentDispenserSetDispenseAmountMessage(FixedPoint2 amount) : BoundUserInterfaceMessage
+{
+    public readonly FixedPoint2 ReagentDispenserDispenseAmount = amount;
+}
+
+[Serializable, NetSerializable]
+public sealed class ReagentDispenserDispenseReagentMessage(ItemStorageLocation storageLocation)
+    : BoundUserInterfaceMessage
+{
+    public readonly ItemStorageLocation StorageLocation = storageLocation;
+}
+
+/// <summary>
+/// Message sent by the user interface to ask the reagent dispenser to eject a container
+/// </summary>
+[Serializable, NetSerializable]
+public sealed class ReagentDispenserEjectContainerMessage(ItemStorageLocation storageLocation)
+    : BoundUserInterfaceMessage
+{
+    public readonly ItemStorageLocation StorageLocation = storageLocation;
+}
+
+[Serializable, NetSerializable]
+public sealed class ReagentDispenserClearContainerSolutionMessage : BoundUserInterfaceMessage;
+
+[Serializable, NetSerializable]
+public sealed class ReagentInventoryItem(
+    ItemStorageLocation storageLocation,
+    string reagentLabel,
+    FixedPoint2 quantity,
+    Color reagentColor)
+{
+    public ItemStorageLocation StorageLocation = storageLocation;
+    public string ReagentLabel = reagentLabel;
+    public FixedPoint2 Quantity = quantity;
+    public Color ReagentColor = reagentColor;
+}
+
+[Serializable, NetSerializable]
+public sealed class ReagentDispenserBoundUserInterfaceState(
```

Comment:
similar concern to https://github.com/space-wizards/space-station-14/pull/38998 wrt using a BUI state instead of compstates, especially now that all the data is in shared

Reply:
I have a commit locally that converts this to component states, but I'm stuck on the same issue mentioned on the partner PR: I can't figure out how to (idiomatically) get the updated states of child entities, since their states haven't been applied yet. This leads to an issue where the solution contents won't be changed for an observing client while another client dispenses stuff.

Reply:
Actually I feel like this does need a second-level relay for solution updates which is... a little annoying. Hrm.